pkg/kstatus/observe/observe.go

```diff
@@ -0,0 +1,309 @@
+	return &StatusObserver{
+		reader: reader,
+		mapper: mapper,
+	}
+}
+
+// StatusObserver provides functionality for polling a cluster for status for a set of resources.
+type StatusObserver struct {
+	reader client.Reader
+	mapper meta.RESTMapper
+}
+
+// EventType is the type that describes the type of an Event that is passed back to the caller
+// as resources in the cluster are being observed.
+type EventType string
+
+const (
+	// ResourceUpdateEvent describes events related to a change in the status of one of the observed resources.
+	ResourceUpdateEvent EventType = "ResourceUpdated"
+	// CompletedEvent signals that all resources have been reconciled and the observer has completed its work. The
+	// event channel will be closed after this event.
+	CompletedEvent EventType = "Completed"
+	// AbortedEvent signals that the observer is shutting down because it has been cancelled. All resources might
+	// not have been reconciled. The event channel will be closed after this event.
+	AbortedEvent EventType = "Aborted"
+	// ErrorEvent signals that the observer has encountered an error that it can not recover from. The observer
+	// is shutting down and the event channel will be closed after this event.
+	ErrorEvent EventType = "Error"
+)
+
+// Event defines that type that is passed back through the event channel to notify the caller of changes
+// as resources are being observed.
+type Event struct {
+	// EventType defines the type of event.
+	EventType EventType
+
+	// AggregateStatus is the collective status for all the resources. It is computed by the
+	// StatusAggregator
+	AggregateStatus status.Status
+
+	// Resource is only available for ResourceUpdateEvents. It includes information about the observed resource,
+	// including the resource status, any errors and the resource itself (as an unstructured).
+	Resource *common.ObservedResource
+
+	// Error is only available for ErrorEvents. It contains the error that caused the observer to
+	// give up.
+	Error error
+}
+
+// Observe will create a new StatusObserverRunner that will poll all the resources provided and report their status
+// back on the event channel returned. The StatusObserverRunner can be cancelled at any time by cancelling the
+// context passed in.
+// If stopOnCompleted is set to true, then the runner will stop observing the resources when the StatusAggregator
+// determines that all resources has been fully reconciled. If this is set to false, the observer will keep running
+// until cancelled.
+// If useCache is set to true, the observer will fetch all resources needed using LIST calls before each polling
+// cycle. The observers responsible for computing status will rely on the cached data. This can dramatically reduce
+// the number of calls against the API server.
+func (s *StatusObserver) Observe(ctx context.Context, resources []wait.ResourceIdentifier, stopOnCompleted bool, useCache bool) <-chan Event {
```

Comment:
nit: consider making `useCache` a field on StatusObserver

Reply:
Since `useCache` is only used within the `Observe` function, I think it makes sense to keep this as a parameter.